### PR TITLE
Adapter adc detection

### PR DESCRIPTION
**Pull request by yangtechen**
Add a vendor command 39 to read adapter ADC value.
- For the latest adapter, it will read ~48k
- For the legacy GQ/NQ/Spencer adapter, it will read ~55k

### CODE REVIEW

File: source/rtos2/Include/RTX_Config.h
Comment (gaborcsapo) [inline source/rtos2/Include/RTX_Config.h]: We don't want to enable this for all other HICs. Can we just set the macro in the stm32h743ii_if yaml?

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h]: I think we need to separate the adc and the adapter_detector further.

The ADC should be initialized as part of each ADC read, because we might use the adc for other purposes in between. The adc doesn't strictly belong to the adapter detector. Epidote has something similar, you can take a look at nlnq_accessory.c

At nest, we also like to prefix our global functions with the file name. So if this file becomes adc_detector, the function names would be like adc_detector_update()

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h]: nit (small thing): the space after the macro's name is inconsistent. Generally the all macro values should be vertically aligned by using tabs.

File: source/board/stm32h743ii.c
Comment (gaborcsapo) [inline source/board/stm32h743ii.c]: a more descriptive name would be udb_extended_features_task, just an idea

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: since your next task is looked at asserts and crash info, I think asserting on errors might be more informative.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c]: 100ms sounds unnecessarily frequent. I think every 3 seconds is perfectly fine.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c]: where is this flag from? if noone would build UDB with this legacy cmsis os, then we don't need these flags and the file is more readable.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c
Comment (gaborcsapo) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c]: we could move all our init code from the prerun_board_config here. As the list of features grow, it's probably better to keep it here than in source/board

File: source/board/stm32h743ii.c
Comment (gaborcsapo) [reply source/board/stm32h743ii.c]: how about adding a flag here so that when we contribute back the code, people can turn off the extended features easily.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c
Comment (gaborcsapo) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c]: it's also nice to keep all the extended feature related code in one folder instead of letting it scatter out.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/DAP_vendor_ex.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/DAP_vendor_ex.h:14]: The "FLEX_HOTPLUG" seems a little strange. How about something more descriptive of what it is, like "READ_UDC_ADAPTER_TYPE_ADC"?

I'm also wondering - is it more helpful to return an ADC value or to return the type of the connected adapter? E.g., if   ~48k is type "1" and ~55k is type "2", then maybe you could have this be READ_UDC_ADAPTER_TYPE and return 0 for nothing connected, 1 for Joker board, 2 for Spencer, etc.

I'd imagine for debugging, it might be nice to have a command for returning the ADC value, but for real use seems to me like a type might be better.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h]: In addition, maybe change the name to AUX_TASK_FLAG_100MS, to be more consistent in the naming.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h]: Would prefer the name aux_task_create for this function

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/aux_task.h]: Does this need to be in the header? I think it can be a static function, local to aux_task.c.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h]: +1 on naming

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c]: Out of curiosity, do you know that the response buffer has this many bytes available to write into without overrun?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c:282]: It seems you put 4 items here. Is this number correct?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c]: I think you want *response++ here, since otherwise your 2 on the next line will overwrite the DAP_OK (the post-increment will happen only after the assignment, so these first two lines will write the same address twice).

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: Related question: do you have a watchdog? If you were to fall into this loop, would the board reset eventually?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: Can you add a comment at the top of this file to describe what it's for?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: nit: inconsistent indentation

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: nit: inconsistent indentation

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: nit: inconsistent placement of braces

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: What does the 10 mean? Is this somewhere that would be suitable to add a constant for this value?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: Will this only ever be called from one task? Do we need to worry about locking?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: nit: indentation. Here, and in the next function.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: It would be more typical to put the UNUSED(hadc) at the beginning of the function rather than at the end.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c]: Can you add a comment to describe what this file is for? I'm not sure what an "aux task" is exactly, so it would be helpful to have an explanation.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c]: What does the 3 mean here? Can you define some constant to give a name for it?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/aux_task.c]: Based on Gabor's comment, I'm guessing these ifdefs will get deleted, but if they stick around, then please flip the logic to ifdef-else instead of ifndef-else - it's a little bit easier to understand that way.

File: source/rtos2/Include/RTX_Config.h
Comment (aaronokano) [inline source/rtos2/Include/RTX_Config.h]: Why is this needed?

Comment (yangtechen) [reply source/rtos2/Include/RTX_Config.h]: We want to put our own logic in another thread, I need to enable Round-Robin so that the scheduler could preempt the main task.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: I would add a TODO comment for now. I will add crash info after I finish the logging.

Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: Not sure if we have a watchdog. I know that we do not turn on ST's WDG HAL driver.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/DAP_vendor_ex.c]: The response buffer is defined by the macro DAP_PACKET_SIZE, which is 64 bytes.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_adc.c]: This is only called in one task now. But yes, I think I should add a lock here to prevent the potential adc operation from  the main task.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h]: Ok. There is an adc interface file `source/hic_hal/adc.h` after merging the upstream so I will move the logic of adc to `source/hic_hal/adc.c` and only keep adapter detector under the `extended_features/`

File: source/rtos2/Include/RTX_Config.h
Comment (aaronokano) [reply source/rtos2/Include/RTX_Config.h]: Does the main task never yield?

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h]: `source/hic_hal/adc.h` provides an interface `adc_read_channel(uint32_t channelGroup, uint32_t channelNumber, uint32_t channelMux)`.
But as far as I know, stm32h7x does not have channel group for ADC.

File: source/rtos2/Include/RTX_Config.h
Comment (yangtechen) [reply source/rtos2/Include/RTX_Config.h]: Yes, if the we left the Round-Robin disabled, the aux_task will never run.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_adc.h]: I think that's not a problem. If there's no channel group supported by this chip, then I think it should be no problem to have your implementation ignore the channelGroup parameter.

File: source/hic_hal/stm32/stm32h743ii/adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/adc.c]: Can you make this ADC_CONVERSION_TIMEOUT_MS, so the unit is clear in the location where it is used?

File: source/hic_hal/stm32/stm32h743ii/adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/adc.c]: I think just adc_value is ok. This looks like it's supposed to be somewhat generic, so not limited to reading the UDC adapter type, is that right?

File: source/hic_hal/stm32/stm32h743ii/adc.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/adc.c]: Should this function multiple tasks attempting an ADC conversion simultaneously? How is this done in other ADC drivers?

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h]: Are there already reserved values for future adapters? Can we go ahead and set them up here ahead of time so that we can return a correct adapter_type_t even if we don't know which specific adapter it is yet?

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h]: Should there be an ADAPTER_NONE value? Having nothing attached and having an unknown adapter attached are two different cases.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h]: Would this be better in a table of some sort to allow easy iteration?

    // Table to map adapter type to min ADC value that can be read for type
    static const uint16_t s_adapter_type_adc_map[] = {
        [ADAPTER_UDB_13_FLEX] = 54000,
        [ADAPTER_UDB_JOKER_FLEX] = 51000,
        ...
    };

Also might be better to put it into the C file where it's used, since I suspect it will only be one file that needs this info.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h]: I think these defines can all be moved into udb_extended_features_task.c where they are used. No need to cause other files to include more definitions than necessary.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c:41]: Don't we read the HW ID before this point? Doesn't the HW ID require the ADC? We should probably read it after initializing the ADC.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h]: Is the 10ms tick a fixed value? If not, is there a way to retrieve the tick rate and define this value in terms of the tick rate?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c:41]: ADC is only used for adapter detector. HW ID is detected by GPIO read.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/udb_extended_features_task.h]: osKernelGetTickFreq() could work

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h]: Good suggestion, see this syntax the first time.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h]: I have no idea if there are reserved value. But I think it is hard to name those adapters, I will just use ADAPTER_UNKNOWN now.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c:41]: Really? I'm not sure if I have the correct schematics, but it seems like proto2 has a voltage divider to express the version and is connected to an ADC input. I guess if we just want to distinguish proto1 and proto2, a GPIO could work (although I don't quite understand how it can work at all in proto1 regardless, since the pin is floating).

I suppose addressing the hardware version reading can be done later, but maybe it would be good to pre-emptively move the HW version read to after ADC init, since at some point, at the very minimum when the next hardware revision after proto2 comes out, it's going to need to use the ADC.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h]: I think it would be fine to just write something like ADAPTER_TYPE_6, ADAPTER_TYPE_7, etc. The reason I'm suggesting this is because at some point in the future, people will be making new adapters and you probably do not want to require a UDB firmware update for each one just to read which type it is. It's better to have a mapping established ahead of time so at least the type value can be read for each one, even if you don't internally have a name for it.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h]: One more suggestion on the adapter type enum - I suggest explicitly listing the value of each enum member so that one can easily look at the enum and know which number corresponds to which type.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h]: I suggest reserving a number outside of the range of normal types for the UNKNOWN and NONE types. You could maybe have ADAPTER_NONE = 0 and ADAPTER_UNKNOWN = 0xFF or something along those lines (not sure what value is most suitable here). This is because at some point you will add more adapter types, and then the numbers that used to mean UNKNOWN and NONE will mean something else. Better to keep them the same, especially if these types will be part of the software interface at some point.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter_detector.h]: Sorry, also meant to add an example of what I meant:

    typedef enum
    {
        ADAPTER_UDB_13_FLEX = 0,
        ADAPTER_JOKER_FLEX  = 1,
        ...
    };

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_detector.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_detector.c:18]: What pin does this correspond to? Can you add a comment for that?

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c:41]: ADC1_2_3_EXTI15_R is just for ADC external trigger and can not be used as ADC input. So what Eric came up with is that we can enable internal weak pull-up register for the GPIO to make proto1 read high and proto2 read low. But you are right that this can only differentiate p1 and p2 and can not be used for future versions.

File: source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/extended_features/include/adapter.h]: I found a pin-mapping for UDB, it seems there are some reserved values https://docs.google.com/spreadsheets/d/1crGH2GwyluorJVSYZoUDuaexZTivegYbg4MbRePD_6s/edit#gid=0

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c
Comment (aaronokano) [reply source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c:41]: How is this going to work for future versions? Are they going to move to a different pin?

This sounds like it's going to be a headache for builds after p2. Anyway, I think this is a discussion for another time.

File: source/hic_hal/stm32/stm32h743ii/adc.c
Comment (yangtechen) [reply source/hic_hal/stm32/stm32h743ii/adc.c]: I will add a mutex to prevent such race condition.

File: source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_detector.c
Comment (aaronokano) [inline source/hic_hal/stm32/stm32h743ii/extended_features/source/adapter_detector.c:41]: Can you add a comment to say that this is an arbitrarily chosen low value between 0 and adapter type 15?

**Comment by aaronokano** [comment]
This looks good to me! Thanks, Yang-te. Looks like Gabor or Eric needs to give the final approval.

**Review by eleenest — DISMISSED** [review]

**Review by gaborcsapo — APPROVED** [review]

**Pull request merged**
